File: bytopic/bitmanipulation/java/group2/SingleNumberII.java
// Given an integer array nums where every element appears three times except for one, which appears exactly once.
// Find the single element and return it.
// You must implement a solution with a linear runtime complexity and use only constant extra space.
// Example:
// Input: nums = [0,1,0,1,0,1,99]
// Output: 99
//
// Tag: 129/150
// Tag: 137/2927, R139/2936 (overall frequency ranking)

class SingleNumberII {
    public static void main(String[] args) {
       int[] nums = {0,1,0,1,0,1,99};
       System.out.println("Single number: " + singleNumber(nums));
    }

    static int singleNumber(int[] nums) {
        int ans = 0;
        for (int i = 0; i < 32; i++) {
            int sum = 0;
            for (int j = 0; j < nums.length; j++) {
                if (((nums[j] >> i) & 1) == 1) {
                    sum++;
                    sum %= 3;
                }
            }
            if (sum != 0) {
                ans |= sum << i;
            }
        }
        return ans;
    }
}

// Method:
// Think about the number in 32 bits and just count how many 1s are there in each bit, and sum %= 3 will clear it once
// it reaches 3. After running for all the numbers for each bit, if we have a 1, then that 1 belongs to the single number,
// we can simply move it back to its spot by doing ans |= sum << i;
//
// This has complexity of O(32n), which is essentially O(n) and very easy to think and implement.
// Als we have a general solution for any times of occurrence. Say all the numbers have 5 times, just do sum %= 5.

File: bytopic/bitmanipulation/python/group2/single_number_ii.py
"""
Given an integer array nums where every element appears three times except for one, which appears exactly once.
Find the single element and return it.
You must implement a solution with a linear runtime complexity and use only constant extra space.

Example:
Input: nums = [0,1,0,1,0,1,99]
Output: 99

Tag: 129/150
Tag: 137/2927, R186/2936 (overall frequency ranking)
"""


def single_number(nums: list[int]) -> int:
    ans = 0
    for i in range(32):
        bit_sum = 0
        for num in nums:
            # Convert the number to two's complement representation to handle large test case
            if num < 0:
                num = num & (2**32-1)
            bit_sum += (num >> i) & 1
        bit_sum %= 3
        ans |= bit_sum << i
    # Convert the result back to two's complement representation if it's negative
    if ans >= 2**31:
        ans -= 2**32
    return ans


def single_number_1(nums: list[int]) -> int:
    ones = 0
    twos = 0

    for num in nums:
        ones ^= (num & ~twos)
        twos ^= (num & ~ones)
    return ones


def main():
    nums = [0, 1, 0, 1, 0, 1, 99]
    print(single_number_1(nums))


if __name__ == "__main__":
    main()
